STACpopulator/extensions/base.py

```diff
@@ -0,0 +1,330 @@
+from datetime import datetime
+import json
+import jsonschema
+import logging
+from typing import Any, Dict, Generic,    TypeVar, Union, cast, Optional
+from pydantic import (BaseModel, create_model, Field, FilePath, field_validator, model_validator, HttpUrl, ConfigDict,
+                      PrivateAttr)
+import pystac
+from pystac.extensions import item_assets
+from pystac.extensions.base import (
+    ExtensionManagementMixin,
+    PropertiesExtension,
+    SummariesExtension,
+)
+from pystac import STACValidationError
+from pystac.extensions.base import S  # generic pystac.STACObject
+from STACpopulator.models import AnyGeometry, GeoJSONPolygon
+from STACpopulator.stac_utils import (
+    ServiceType,
+    ncattrs_to_bbox,
+    ncattrs_to_geometry,
+)
+import types
+from pystac.extensions.datacube import DatacubeExtension
+from STACpopulator.extensions.datacube import DataCubeHelper
+from STACpopulator.extensions.thredds import THREDDSExtension, THREDDSHelper
+
+
+
+T = TypeVar("T", pystac.Collection, pystac.Item, pystac.Asset, item_assets.AssetDefinition)
+
+LOGGER = logging.getLogger(__name__)
+"""
+# Context
+
+What we have:
+  - `Loader`, which returns attributes.
+  - An external json schema describing a subset of the attributes returned by the Loader. This schema might preclude
+  additional properties, so it cannot be applied wholesale to the Loader's output. (maybe overkill since not a lot of schemas can be found in the wild...)
+  - `data model` describing the content we want included in the catalog. It includes a subset of the schema properties,
+  as well as additional attributes desired by the catalog admins.
+
+Desiderata:
+  - Not having to replicate existing validation logic in the schema
+  - Not having to create a modified schema
+  - Being able to supplement the schema validation by pydantic validation logic
+  - Streamline the creation of new data models (reduce boilerplate, allow subclassing)
+  - Developer-friendly validation error messages
+
+
+How-to:
+  - Instructions to create basic datamodel from schema (codegen)
+  
+  
+  
+"""
+class DataModel(BaseModel):
+    """Base class for dataset properties going into the catalog.
+
+    Subclass this with attributes.
+    """
+    # Ideally, the catalog properties would be described by a jsonschema.
+    _prefix: str = PrivateAttr()
+
+    # URI of the json schema to validate against.
+    _schema_uri: FilePath = PrivateAttr(None)
+
+    # List of properties not meant to be validated by json schema.
+    _schema_exclude: list[str] = PrivateAttr([])
```

Comment:
I don't think so, because those were fields I wanted to exclude from the schema validation, but not from the model dump. I was thinking of a case where the schema is strictly prohibiting extra attributes, but I realize this might be a very edgy corner case.